x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -125,14 +129,51 @@ public void execute(
         );
     }
 
+    /**
+     * Execute an analyzed plan. Most code should prefer calling {@link #execute} but
+     * this is public for testing. See {@link Phased} for the sequence of operations.
+     */
     public void executeAnalyzedPlan(
         EsqlQueryRequest request,
         BiConsumer<PhysicalPlan, ActionListener<Result>> runPhase,
         LogicalPlan analyzedPlan,
         ActionListener<Result> listener
     ) {
-        // TODO phased execution lands here.
-        runPhase.accept(logicalPlanToPhysicalPlan(analyzedPlan, request), listener);
+        LogicalPlan firstPhase = Phased.extractFirstPhase(analyzedPlan);
```

Comment:
I felt like doing it with a LogicalPlan was right because each child query is valid and can get optimized.

But it does make sense to do the logical optimization first.

Reply:
I think the current way is not wrong, it just prevents phased plan nodes to ever be subject to logical optimizations; and the first phase can never profit from knowing that a part of the result it generates are irrelevant, because the first phase gets currently optimized completely independently.

Since phased plan nodes are generally expensive, I think we'll want and have to optimize them.

Reply:
We're going to save this stuff for the follow up - it's filed under " Run logical plan optimization before splitting the phases. "